### PR TITLE
Convert numpy scalars to python objects before dumping to yaml

### PR DESCRIPTION
**Pull request by Bluesy1**
Slightly different approach as what was described on mattermost, but I tested the mattermost suggestion and it didn't seem to work.

see the one test case that changed for evidence this works.

### CODE REVIEW

File: src/problem_bank_scripts/problem_bank_scripts.py [hunk 1]
```diff
@@ -80,6 +80,11 @@ def defdict_to_dict(defdict, finaldict):
                     v[k2] = dict(sorted(v2.items(), key=lambda i: i[0]))
                 
             finaldict[k] = v
+        elif hasattr(v, "dtype"):
+            try:
+                finaldict[k] = v.item()
+            except:
```
Comment (firasm) [inline src/problem_bank_scripts/problem_bank_scripts.py:86]:
```suggestion
            except ConstructorError:
```

This should be a ConstructorError right ([ref](https://github.com/open-resources/stats_bank/actions/runs/10519721270/job/29147553222#step:6:35))? Generally not good practice to catch all errors

File: src/problem_bank_scripts/problem_bank_scripts.py [hunk 2]
```diff
@@ -80,6 +80,11 @@ def defdict_to_dict(defdict, finaldict):
                     v[k2] = dict(sorted(v2.items(), key=lambda i: i[0]))
                 
             finaldict[k] = v
+        elif hasattr(v, "dtype"):
```
Comment (firasm) [inline src/problem_bank_scripts/problem_bank_scripts.py:83]: Can we add a comment for why this happens so we can have a hope of debugging if anything goes wrong?

```suggestion
        # Somewhere around Aug 2024, something changed in how objects were serialized and anything stored as a `numpy` object wasn't correctly parsed by `pyyaml`
        # See here for our history on this: (https://moosvilab.ok.ubc.ca/moosvilab/pl/zfp63dh6hjy5zdkzemzmtzqffr)
        # And the jupyterbook PR where this was also addressed: https://github.com/jupyter-book/jupyter-book/pull/2167

        elif hasattr(v, "dtype"):
```

**Review by firasm — APPROVED** [review]
Couple of suggestions to add (if you think they're appropriate) but this can be merged in after

File: src/problem_bank_scripts/problem_bank_scripts.py (hunk 1 above)
Comment (Bluesy1) [reply src/problem_bank_scripts/problem_bank_scripts.py:90]: > This should be a ConstructorError right ([ref](https://github.com/open-resources/stats_bank/actions/runs/10519721270/job/29147553222#step:6:35))? Generally not good practice to catch all errors

pyyaml's `ConstructorError` would be the wrong error to use here.

A middle ground is to catch an `Exception`, such that we just don't convert if the `.item()` vall fails - I'd rather to bail out over failing catastrophically when converting the defaultdict to a standard dict if a non numpy object that has a `dtype` attribute gets passed to a data dictionary - I don't know where that could happen, but its possible.

```suggestion
            except Exception:
```

**Pull request merged**